### PR TITLE
[pack] Reacting to ScriptHost configuration changes for AzureStorageProvider

### PR DESCRIPTION
**Pull request by karshinlin**
### Issue describing the changes in this PR

Resolves issue with custom configuration sources set by App not accessible by AzureStorageProvider. Added an event handler for AzureStorageProvider to update the used IConfiguration with that of the active JobHost.

### Pull request checklist

* [X] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script/StorageProvider/IAzureStorageProvider.cs
Comment (brettsam) [inline src/WebJobs.Script/StorageProvider/IAzureStorageProvider.cs:19]: Should this be a required param, or is sending null okay in some cases? Looking at FunctionsSyncManager, for instance...

File: src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs
Comment (brettsam) [inline src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs:37]: This means that no WebHost-level service should be caching any of the blob clients created by this, right? B/c they can change whenever there's a new host? Which services were we worried about for that?

Not sure if I like that long-term, but as a fast, pragmatic fix, it should be okay.

File: src/WebJobs.Script/StorageProvider/IAzureStorageProvider.cs
Comment (karshinlin) [reply src/WebJobs.Script/StorageProvider/IAzureStorageProvider.cs:19]: So in most cases where IAzureStorageProvider is used, the IConfiguration will update with that of the active JobHost -- meaning this parameter isn't needed and null is okay. There's a case when the active JobHost isn't started yet but we need it's configuration (deciding between blob lock and in-memory lock manager during ScriptHost service initialization) when we need to pass the configuration as a parameter. 

FunctionsSyncManager is in the context of a controller which already has the right IConfiguration I believe (but the added eventhandler logic takes care of this too). FunctionsSyncManager relies on the BlobStorageSecretsRepository through dependency injection though, which is still using the WebHost configuration and has trouble accessing storage when customers don't set AzureWebJobsStorage AppSetting (in previous versions too).

File: src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs
Comment (karshinlin) [reply src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs:37]: Services that use IAzureStorageProvider:
ScriptHost:
- BlobLeaseDistributedLockManager
- AzureScheduleStorageMonitor (caches blobContainerClient)

WebHost:
- BlobChangeAnalysisStateProvider (about to be removed)
- FunctionsSyncManager (caches blobContainerClient)
- BlobStorageSecretsRepository (caches blobClient)

We do need to consider the effects of caching a client if the IConfiguration changes. Do ScriptHost services get disposed and replaced when ScriptHost gets restarted? When does the outer WebHost get restarted?

**Comment by karshinlin** [comment]
Note that a custom configuration sourcing AzureWebJobsStorage will still have problems with BlobStorageSecretsRepository (used by FunctionsSyncManager). This problem exists in 3.0.15584.0 and this change.

File: src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs
Comment (brettsam) [reply src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs:37]: Yes, ScriptHost services are disposed and re-created when it restarts. The outer WebHost only disposes when the process exits.

**Review by brettsam — DISMISSED** [review]
I'm good with this now, knowing that we'll have some refactoring coming in the future that means we can potentially revisit the fragility of how this works at WebHost/JobHost level.

Still want @fabiocav to weigh in

**Comment by Hazhzeng** [comment]
HotFix branch **release/3.0-hotfix** is now even with **release/3.0**.
Please create a PR targeting **release/3.0-hotfix** as well once this PR got merged.

File: src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs
Comment (fabiocav) [inline src/WebJobs.Script/StorageProvider/AzureStorageProvider.cs:83]: This might benefit from deferral. No change required for now, but wanted to flag.

File: src/WebJobs.Script/StorageProvider/IAzureStorageProvider.cs
Comment (fabiocav) [reply src/WebJobs.Script/StorageProvider/IAzureStorageProvider.cs:19]: From the `FunctionsSyncManager` you should aways have access to `IConfiguration`, but that could come from different scopes depending on the scenario, but this simply means the the right instance will be passed to you automatically. 

Being defensive and correctly logging unexpected conditions (which is what Brett is concerned about) is advisable.

File: src/WebJobs.Script/StorageProvider/StorageClientProvider.cs
Comment (fabiocav) [inline src/WebJobs.Script/StorageProvider/StorageClientProvider.cs:120]: This code should be more defensive. Even with the check that exists in `GetWebJobsConnectionStringSection` (actually indirectly in the method it ends up calling), not performing a `null` check here makes this a bit fragile. Please update to perform a check.

**Review by fabiocav — DISMISSED** [review]
A few nits, but not blocking this for now. Just wanted to capture and we should address ASAP.

**Comment by alrod** [comment]
@karshinlin,  Can you please add the fix to release notes?

**Review by alrod — APPROVED** [review]

**Comment by eranjbarSuncor** [comment]
Will there be a similar fix for the Python function? Noticed that this version isn't available/can't be referenced by the Python function apps

**Pull request merged**
